New semantic analyzer: support named tuples

ilevkivskyi opened the pull request:
Fixes #6284 
Fixes #6285 

This doesn't add support for recursive definitions, but all the forward reference patterns that are supported in old semantic analyzer, should be also supported.

Some notes:
* I moved one function to `semanal_namedtuple.py`
* I never create new `TypeInfo`s in multi-iteration scenario
* I add a dummy type variable `_NT` to the bodies (otherwise type variable scope breaks), but it is probably fine because names starting with underscores are anyway prohibited in named tuples.

JukkaL commented on mypy/newsemanal/semanal_namedtuple.py:
```diff
@@ -403,8 +431,42 @@ def make_init_arg(var: Var) -> Argument:
 
+    @contextmanager
+    def save_namedtuple_body(self, named_tuple_info: TypeInfo) -> Iterator[None]:
+        """Analyze body of class-based named tuple if the NamedTuple base class is present.
```
The docstring doesn't seem right.

JukkaL commented on mypy/newsemanal/semanal_namedtuple.py:
```diff
@@ -98,7 +118,8 @@ def check_namedtuple_classdef(
                     analyzed = self.api.anal_type(stmt.type)
-                    assert analyzed is not None  # TODO: Handle None values
+                    if analyzed is None:
+                        return None
```
Add comment about this (something is incomplete).

JukkaL commented on mypy/newsemanal/semanal_namedtuple.py:
```diff
@@ -35,35 +35,55 @@
-                        return info
-        return None
+                    if defn.info:
+                        # Don't reprocess everything, potentially deferred methods will be
```
Grammar nit: Run-on sentence, use period instead of comma.

JukkaL commented on mypy/newsemanal/semanal.py:
```diff
@@ -867,6 +867,17 @@ def analyze_class(self, defn: ClassDef) -> None:
 
+        is_named_tuple, info = self.named_tuple_analyzer.analyze_namedtuple_classdef(defn)
+        if is_named_tuple:
+            if info is None:
```
I'd suggest moving the new code to a separate method, as the method is already quite long.

JukkaL commented on mypy/newsemanal/semanal.py:
```diff
@@ -1293,7 +1264,7 @@ def configure_base_classes(self,
         for base, base_expr in bases:
             if isinstance(base, TupleType):
-                if info.tuple_type:
+                if info.tuple_type and info.tuple_type != base:
```
Is there any case where `info.tuple_type` and `base` would be equal? I'd expect that at least the fallbacks would be different. Also, we usually use `is_same_type` instead of equality.

JukkaL commented on mypy/newsemanal/semanal.py:
```diff
@@ -1882,7 +1853,12 @@ def visit_assignment_stmt(self, s: AssignmentStmt) -> None:
         self.process_typevar_declaration(s)
-        self.named_tuple_analyzer.process_namedtuple_definition(s, self.is_func_scope())
+        if isinstance(s.rvalue, CallExpr) and isinstance(s.rvalue.analyzed, NamedTupleExpr):
+            existing_info = s.rvalue.analyzed.info  # type: Optional[TypeInfo]
```
I think that it would be better to do nothing on subsequent iterations if we've already successfully analyzed a named tuple definition. Analysis should only be successful if everything can be bound, and in that case the second iteration should produce the same result. However, in case there is a bug that causes the first iteration to produce an incomplete or invalid type, the second iteration could hide this bug in many cases, making the bug harder to find. Additionally, the logic would be a bit simpler.

We can't use this with normal class definitions though, since base types can be incomplete after the first iteration.

JukkaL commented on mypy/newsemanal/semanal_namedtuple.py:
```diff
@@ -35,35 +35,55 @@
+                        return True, defn.info
+                    result = self.check_namedtuple_classdef(defn)
+                    if result is None:
+                        return True, None
```
Add short comment describing what this means.

JukkaL commented on mypy/newsemanal/semanal_namedtuple.py:
```diff
@@ -35,35 +35,55 @@
+                    defn.analyzed.line = defn.line
+                    defn.analyzed.column = defn.column
+                    return True, info
+        return False, None
```
Add short comment here.

JukkaL commented on mypy/newsemanal/semanal_namedtuple.py:
```diff
@@ -35,35 +35,55 @@
+                    defn.analyzed = NamedTupleExpr(info, is_typed=True)
+                    defn.analyzed.line = defn.line
+                    defn.analyzed.column = defn.column
+                    return True, info
```
Add short comment here.

JukkaL commented on mypy/newsemanal/semanal_namedtuple.py:
```diff
@@ -115,16 +136,21 @@ def check_namedtuple_classdef(
 
-        Assume that s has already been successfully analyzed.
+        Assume that s has already been successfully analyzed. Use existing_info
+        if given instead of creating a new one.
```
As discussed above, I think that it would be better to avoid calling this method if this has been successfully analyzed before.

JukkaL commented on mypy/newsemanal/semanal_namedtuple.py:
```diff
@@ -403,8 +431,42 @@ def make_init_arg(var: Var) -> Argument:
+        """Analyze body of class-based named tuple if the NamedTuple base class is present.
+
+        Temporarily clear the names dict so we don't get errors about duplicate names
+        that were already set in build_namedtuple_typeinfo.
```
Mention why these shouldn't produce errors.

ilevkivskyi commented:
@JukkaL Thanks for review!

Because you merged your PR first, it was necessary to update my implementation logic for assignment based named tuples (basically it is much closer to class based one now). I however discovered couple small bugs:
* Wrong full name was set for named tuples nested in methods (this area around names with `@line` is still dark and will need some clean-up later).
* Processing of functions/methods were stopping to soon.

I fixed them and added more tests.

JukkaL commented on mypy/newsemanal/semanal.py:
```diff
@@ -1880,17 +1888,32 @@ def visit_assignment_stmt(self, s: AssignmentStmt) -> None:
+            return
+        # Yes, it's a valid namedtuple, but defer if it is not ready.
+        if not info:
+            self.add_symbol(name, PlaceholderNode(self.qualified_name(name), lvalue, True),
```
`mark_incomplete` adds a `PlaceholderNode`. The logic is a bit delicate so it's important that all additions go through the same implementation.

JukkaL commented on mypy/newsemanal/semanal_main.py:
```diff
@@ -123,7 +123,7 @@ def process_top_level_function(analyzer: 'NewSemanticAnalyzer',
     while deferred and more_iterations:
         iteration += 1
-        if not incomplete or iteration == MAX_ITERATIONS:
+        if not (deferred or incomplete) or iteration == MAX_ITERATIONS:
```
A namespace can be deferred while it's incomplete. I think that we should also remove the namespace from incomplete namespaces if it's complete. Now it looks like that we sometimes don't remove the namespace from `incomplete_namespaces` even though we probably should.

JukkaL commented on test-data/unit/check-newsemanal.test:
```diff
@@ -805,6 +825,61 @@ Out = NamedTuple('Out', [('x', In), ('y', Other)])
 In = NamedTuple('In', [('s', str), ('t', Other)])
 class Other: pass
 
+[case testNewAnalyzerNamedTupleMixed1]
```
Is there a specific reason for adding these tests? It will be impractical to test all combinations of features anyway, as the number of test cases will grow too quickly. For example, is this more useful than a mixed named tuple vs. typed dict test case? It may be better to only test combinations that are known to be risky or have been broken previously.

ilevkivskyi replied:
Yes, these two tests cases were broken, but then the breakage had the same cause as for assignment based vs assignment based, so it is probably OK to remove them.

JukkaL commented on mypy/newsemanal/semanal.py:
```diff
@@ -1214,7 +1214,11 @@ def prepare_class_def(self, defn: ClassDef, info: Optional[TypeInfo] = None) ->
-                defn.info._fullname = self.cur_mod_id + '.' + local_name
+                if defn.info.is_named_tuple:
+                    # Module already correctly set for named tuples.
+                    defn.info._fullname += '@' + str(defn.line)
```
It's unclear why we need this special case. Even if the module was already correctly set, if it's always `self.cur_mod_id`, I think that it's better to avoid this if statement if it doesn't change behavior.

ilevkivskyi replied:
As discussed elsewhere this is (unfortunately) actually needed.
